### PR TITLE
add Gemini as Vision  (OCR) provider

### PR DESCRIPTION
**Pull request by its-a-unixsystem**
**Summary**

Adds Google Gemini as a vision LLM provider for OCR, supporting both image and native PDF processing

Gemini has native PDF handling so both pdf and whole_pdf OCR modes can be used directly.

**Configuration**

```
  VISION_LLM_PROVIDER=googleai
  VISION_LLM_MODEL=gemini-2.5-flash
  GOOGLEAI_API_KEY=your-api-key
  # Optional: thinking budget for reasoning models
  GOOGLEAI_THINKING_BUDGET=16384
  # can be set to image, pdf or whole_pdf
  OCR_PROCESS_MODE=whole_pdf
```

**Changes**

  - `ocr/google_ai_client.go` : New GoogleAIProvider implementing the llms.Model interface with support for text, binary, and data URL image parts
  - `ocr/google_ai_client_test.go` : tests for the new provider
  - `ocr/llm_provider.go` : Adding googleai to the vision provider factory; add MIME detection to route PDFs vs images to the correct content part format (can be debated if this is the right way to go, but I wanted to keep the code short)
  - `ocr/provider.go` : Add GoogleAIAPIKey and GoogleAIThinkingBudget to Config
  - `main.go` : Add googleai case to createVisionLLM(), extend validateOCRProviderModeCompatibility() to allow pdf/whole_pdf modes with googleai

## Summary by CodeRabbit

* **New Features**
  * Added Google AI (Gemini) support for language and vision tasks, with configurable API key and optional "thinking budget" for requests.

* **Improvements**
  * Enhanced multimodal handling: better image vs PDF detection, expanded provider compatibility and rate-limited request paths.

* **Tests**
  * Added comprehensive tests for provider initialization, options, content edge cases, and cleanup.

### CODE REVIEW

File: ocr/google_ai_client.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+		return "", fmt.Errorf("googleai GenerateContent API returned empty response")
+	}
+
+	candidate := resp.Candidates[0]
+	if candidate.Content == nil || len(candidate.Content.Parts) == 0 {
+		return "", fmt.Errorf("googleai GenerateContent API returned a candidate with no content")
+	}
+
+	return candidate.Content.Parts[0].Text, nil
+}
```
[the remark this answers is not in the document]
Comment (its-a-unixsystem) [reply ocr/google_ai_client.go:74]: addressed in 1df32ee

File: ocr/google_ai_client.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	candidate := resp.Candidates[0]
+	if candidate.Content == nil || len(candidate.Content.Parts) == 0 {
+		return nil, fmt.Errorf("googleai GenerateContent API returned a candidate with no content")
+	}
+
+	// Concatenate text parts
+	var sb strings.Builder
+	for _, part := range candidate.Content.Parts {
+		sb.WriteString(part.Text)
+	}
```
[the remark this answers is not in the document]
Comment (its-a-unixsystem) [reply ocr/google_ai_client.go]: addressed in https://github.com/icereed/paperless-gpt/commit/1df32ee3424bfcd0f695e6856f431c9816a72ab4

**Comment by its-a-unixsystem** [comment]
On the outside diff range comment: I already fixed both of these in `createVisionLLM` (rate limiting + using the global for thinking budget). The `createLLM` function is pre-existing code outside this PR's diff.

File: ocr/google_ai_client.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+	return &llms.ContentResponse{
+		Choices: []*llms.ContentChoice{
+			{
+				Content: sb.String(),
+			},
+		},
+	}, nil
+}
```
[the remark this answers is not in the document]
Comment (its-a-unixsystem) [reply ocr/google_ai_client.go:192]: adressed in https://github.com/icereed/paperless-gpt/pull/881/commits/34179671c8af3d28d151882bce1b6412972a847a

**Comment by phaset** [comment]
Thank you for this PR. I wrote about this issue in the discussions here: https://github.com/icereed/paperless-gpt/discussions/867

There remains the langchain challenge of different thinking variables for different versions of Gemini-[x]-flash models: https://ai.google.dev/gemini-api/docs/thinking

**Comment by its-a-unixsystem** [comment]
Yes, Gemini 3's  `thinkingLevel` is not implemented, it is backwards compatible with `thinkingBudget` though, which you can set.
Once GA is out, one could add a switch, but tbh - I would just leave `GOOGLEAI_THINKING_BUDGET` unset and Google uses the default thinking levels/budgets. I haven't seen any quality differences.

I added it as it is available to the other modes and is the only Google setting that is implemented with LangChain (no `withTemperature` or `WithMaxTokens`).

[Gemini 3 is the best OCR available right](https://arena.ai/leaderboard/vision/ocr) now (no need to use pro) - it even managed to decipher my doctors handwritten note.

PS: if you want to test you can just clone my tree - I also have experimental code that does all in one go (send PDF/image -> get fields back, much much faster) https://github.com/its-a-unixsystem/paperless-gpt/tree/feat/oneshot I currently test it and need to clean the code before a PR, also this needs to get merged first.

**Comment by phaset** [comment]
> PS: if you want to test you can just clone my tree - I also have experimental code that does all in one go (send PDF/image -> get fields back, much much faster) https://github.com/its-a-unixsystem/paperless-gpt/tree/feat/oneshot I currently test it and need to clean the code before a PR, also this needs to get merged first.

I might have to try that. Presently I'm doing in the way I've outlined here: https://github.com/icereed/paperless-gpt/discussions/879

**Comment by rgvdds** [comment]
I'm glad to see this! Looking forward to getting it merged and being able to use Gemini for whole_pdf processing!

**Comment by its-a-unixsystem** [comment]
As per https://github.com/icereed/paperless-gpt/discussions/851 the maintainer doesn't have enough time for working a lot on this repo and is looking for a volunteer.
I will maintain this PR, but not create more.

I am rather thinking of creating a more lightweight extension to built upon [Feature: Paperless AI#10319](https://github.com/paperless-ngx/paperless-ngx/pull/10319 )  a AI integration for the upcoming for paperless ngx v3 release.

**Comment by phaset** [comment]
> [Gemini 3 is the best OCR available right](https://arena.ai/leaderboard/vision/ocr) now (no need to use pro) - it even managed to decipher my doctors handwritten note.
> 
> PS: if you want to test you can just clone my tree - I also have experimental code that does all in one go (send PDF/image -> get fields back, much much faster) https://github.com/its-a-unixsystem/paperless-gpt/tree/feat/oneshot I currently test it and need to clean the code before a PR, also this needs to get merged first.

Quick question, so I don't misunderstand. If I use Gemini-3-Flash, then I do not receive back a PDF with the OCR text, correct? Your patch works perfectly, with the OCR tag, and entering the OCR to Postgres, but it doesn't return the PDF with updated text. The main branch documentation indicates this as the default behavior for LLMs OCR, but I wanted to double check.

Thanks for the excellent update!

**Comment by its-a-unixsystem** [comment]
Correct. paperless ngx still does the local OCR process (ideally make it short,  you can't switch it off) paperless-gpt does only replace the fields - for OCR the content field.
That can't really be avoided as you can't change the OCR process for paperless ngx (it always uses OCRmyPDF) nor replace the original PDF via the API. that changes with [paperless ngx 3.0](https://github.com/paperless-ngx/paperless-ngx/milestone/115) (nearly there) which will have an [extension of the API that allows to upload a new version of the original doc/PDF while keeping the old one](https://github.com/paperless-ngx/paperless-ngx/pull/12061).

```mermaid
sequenceDiagram
    participant User as User / Scanner
    participant Paperless as Paperless-ngx
    participant Tesseract as Tesseract (OCR)
    participant GPT as Paperless-GPT
    participant AI as AI Model (OpenAI / Ollama)

    User->>Paperless: 1. Upload Document (PDF/Image)
    
    rect rgb(0, 80, 160)
        Note right of Paperless: OCR Phase
        Paperless->>Tesseract: 2. Send image/pdf for text extraction
        Tesseract-->>Paperless: 3. Return extracted text
        Paperless->>Paperless: 4. Save document & raw text
    end

    rect rgb(160, 60, 0)
        Note right of Paperless: AI Processing Phase
        Paperless->>GPT: 5. Trigger post-consume script / Webhook
        GPT->>Paperless: 6. Fetch document content via API
        Paperless-->>GPT: 7. Return PDF & meta data
        
        GPT->>AI: 8a Send text + prompt correspondent
        AI-->>GPT: 8a Return structured metadata (JSON)
        GPT->>AI: 8b Send text + prompt custom field
        AI-->>GPT: 8b Return structured metadata (JSON)
        GPT->>AI: 8c Send text + prompt created
        AI-->>GPT: 8c Return structured metadata (JSON)
        GPT->>AI: 8d Send text + prompt document type
        AI-->>GPT: 8d. Return structured metadata (JSON)
        GPT->>AI: 8e Send text + prompt tag
        AI-->>GPT: 8e. Return structured metadata (JSON)
        GPT->>AI: 8f Send text + prompt title
        AI-->>GPT: 8f. Return structured metadata (JSON)
        GPT->>AI: 8g. (in case of OCR) Send PDF + prompt (Ask for content)
        AI-->>GPT: 9. Return content as structured metadata (JSON) 
        
        GPT->>Paperless: 10. Update document (Apply tags, correspondent, type) via API
    end
    
    Paperless-->>User: 11. Document fully categorized and searchable
```
The 3.0 AI enhancement will be limited (OpenAI, no OCR via LLM), I would probably do future patches against paperless ngx directly - depending if they are willing to receive them.
